conan/tools/cmake/utils.py

```diff
@@ -47,9 +47,18 @@ def get_generator(conanfile):
         return os.environ["CONAN_CMAKE_GENERATOR"]
 
     compiler = conanfile.settings.get_safe("compiler")
+    compiler_version = conanfile.settings.get_safe("compiler.version")
+    if compiler == "msvc":
+        version = compiler_version[:4]  # Remove the latest version number 19.1X if existing
+        _visuals = {'19.0': '14 2015',
+                    '19.1': '15 2017',
+                    '19.2': '16 2019'}[version]
```

Comment:
Yes, probably we need not only vswhere, but also some easy way to configure it, because if vswhere finds VS2017 and VS2019, and both have the v140 toolset for older versions, which one will it pick? Seems a task for the ``config`` feature, where this can be explicit. And if not explicit, then ``vswhere``, but as the explicit configuration is necessary anyway, I would start with it, then follow with vswhere one.

Reply:
Does it make any difference? The tools (compiler, linker,...) will be exactly the same, won't they? Can you have more than one instance of the same MSVC installed and detected by `vswhere`? Maybe CMake generates something different (VS migrations), but under the hood everything should be exactly the same... I'd say that in this case the generated binary will be exactly the same. Maybe the config is not needed, we can use it if present, otherwise we can choose any at random (or the latest one). cc/ @SSE4

Reply:
that might be different instances of compilers for VS2019 and VS2017 with the same toolset. they probably might have the same version, but I think they installed and updated intependently.

Reply:
I don't think the approach of choosing one at random is a good one. Besides, asking information about the installed toolsets is not possible in vswhere: https://github.com/Microsoft/vswhere/issues/151

Reply:
This is what vswhere does (from that same link): _"[vswhere] finds the installation path (and other basic information) of a version of Visual Studio that meets basic requirements"_ (https://github.com/conan-io/conan/issues/3573#issuecomment-522517346):

```
vswhere [-latest] -requires <component-toolset>
```

The toolset is a requirement, vswhere returns the VS version that satisfies that requirement (VS 2017, VS 2019,...) which is the input we need for the CMake generator.

Reply:
Yeah, I have tried that. The toolset component name that you pass to the ``--requires`` argument is the same (``Microsoft.VisualStudio.Component.VC.Tools.x86.x64``) , and it will map to the default toolset of the different VS versions, not the specific one that you want to discover. So far I have not been able to know which toolsets contain each VS versions, I don't know if it is possible at all.

Reply:
Each toolset has a different component name: 

>   
>   * MSVC `v141`: `Microsoft.VisualStudio.Component.VC.v141.x86.x64`
>   * MSVC `v142` (14.20): `Microsoft.VisualStudio.Component.VC.14.20.x86.x64`
>   * MSVC `v142` (14.21): `Microsoft.VisualStudio.Component.VC.14.21.x86.x64`
>   * `v140`: `Microsoft.VisualStudio.Component.VC.140`
>   * ...
>   
>   It could be easier, but at least [it is documented](https://docs.microsoft.com/es-es/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2019).

Reply:
I have tried them too, but don't seem to work as expected, or at least they are trickier than they look. I didn't succeed in identifying the toolsets I have installed in my different VS versions. I will try again.

Reply:
Results differ from documented (or at least documentation is not very clear).
As in any case, it is necessary to allow configuring the VS/MSBuild instance that is being used to build with a compiler version, let's do the configuration part first, and then we will try to automate to define it automatically.

```
INSTALLED:
    VS 2019: v14.28, v14.16
    VS 2017: v14.16, v14.00
DETECTED
    Microsoft.VisualStudio.Component.VC.Tools.x86.x64 => [VisualStudio.15, VisualStudio.16]
    
    Microsoft.VisualStudio.Component.VC.140 => VisualStudio.15
    Microsoft.VisualStudio.Component.VC.141 => []
    Microsoft.VisualStudio.Component.VC.142 => []

    Microsoft.VisualStudio.Component.VC.14.20 => []
    Microsoft.VisualStudio.Component.VC.14.28 => []
    Microsoft.VisualStudio.Component.VC.14.16 => []

    Microsoft.VisualStudio.Component.VC.14.20.x86.x64 => []
    Microsoft.VisualStudio.Component.VC.14.28.x86.x64 => []
    Microsoft.VisualStudio.Component.VC.14.16.x86.x64 => []

    Microsoft.VisualStudio.Component.VC.v140.x86.x64 => []
    Microsoft.VisualStudio.Component.VC.v141.x86.x64 => VisualStudio.16
    Microsoft.VisualStudio.Component.VC.v142.x86.x64 => []
```